### ISSUE TITLE
[Easy] Order key bindings alphabetically

### ISSUE THREAD

**Issue by MarcoIeni** [VSpaceCode/VSpaceCode#121]
## Problem

Right now key bindings in package.json are not ordered alphabetically.
I am talking about this section of the file:

https://github.com/VSpaceCode/VSpaceCode/blob/17a14cf93ea1f7659ac061524ab5c5741be8daa4/package.json#L131

## Solution

You have to order them alphabetically.
In order to not overlap with each other, comment with the letter(s) you will work on.
For example you can say:

> Hi! I want to work on b (buffers) and c (comments)

**Comment by gnomesoup** [comment]
Hello. I want to work on b (buffers) and c (comments)

**Comment by gnomesoup** [comment]
Except c (comments) is already in order. This is my first pull request so if it goes wall I'll take on more.

**Comment by stevenguh** [comment]
Thank you for contributing. I skim through it, it looks great. I am just trying to add to this discussion where spacemacs order it by non menu keys first then alphabetically. Should we do that too?
<img width="884" alt="image" src="https://user-images.githubusercontent.com/3455662/96390311-e91a0680-1168-11eb-97e1-f1569842d664.png">

**Comment by gnomesoup** [comment]
Great point! I missed that detail. I agree that they should be at the end. For me this alphabetizing and grouping all help me discover the binding that I do not know by heart (which is a lot still in vspacecode).

Do we also want to continue with the "..." for menu items or mimic the "+" prefix convention of spacemacs. Personally, my eye catches the "..." better. 

The highlighted colors for menu items sure is nice. I assume that would require vscode changes.

**Comment by stevenguh** [comment]
Glad you discover something new :)

> Do we also want to continue with the "..." for menu items or mimic the "+" prefix convention of spacemacs. Personally, my eye catches the "..." better.

'...' is used because it is commonly used in OS to indicate that has addition menu than '+' which I have only seen in spacemacs. I am fine either way. But I suggest we table it off until someone opens an issue, and let's focus on the order issue :)

> The highlighted colors for menu items sure is nice. I assume that would require vscode changes.

yeah, I can't think of a way to achieve color. However, I just realized we can use icon on the menu items. They are listed on https://code.visualstudio.com/api/references/icons-in-labels#icon-listing

**Comment by macintacos** [comment]
Btw, unless I'm misunderstanding something, this issue could be handled relatively easily by the "Sort JSON array" extension: https://github.com/fvclaus/vsc-sort-json-array

For the most part, this should be pretty straightforward, and I've used it a few times in my own projects and have some custom sorts that should make things sort the way we expect. I don't mean to detract from the efforts already being made in the community, but when the most recent PR for @gnomesoup is merged, can I just quickly run through all the arrays and open a PR with these changes?

**Comment by MarcoIeni** [comment]
> I don't mean to detract from the efforts already being made in the community, but when the most recent PR for @gnomesoup is merged, can I just quickly run through all the arrays and open a PR with these changes?

If you know an automatic way to do it, sure!

**Comment by stevenguh** [comment]
I also don't want people to step on each other's toes. Let's merge #128 first.

As far as automatic sorting, that sounds great. I ideally would still keep the current key order for each json object (i.e. key, name, type, command, etc).

**Comment by macintacos** [comment]
@stevenguh yep, the order will remain the same - it just sorts the json objects in an array by a given key-value.

**Comment by gnomesoup** [comment]
@macintacos will we need to go in afterwards to get the menu items at the bottom of the list? The grouping aren't entirely alphabetical either since all the lowercase bindings get grouped before the uppercase bindings.

**Comment by stevenguh** [comment]
FYI, #128 is merged, so we can figure out how we can move forward :)

**Comment by macintacos** [comment]
The sorting can be customized. I can make it so that we compare the name first, and if it doesn't have "..." appended at the end, we assume it's an action and place it first, otherwise we place it last (alphabetically). You just have to specify a comparison function.

**Comment by macintacos** [comment]
Actually better yet, we could just put the "type": "bindings" after "type": "command", which is almost certainly more robust than a string manipulation like I just said about 😄

**Comment by stevenguh** [comment]
@macintacos Do you wanna take a stab at it? 🙂 I think the order is
- first by type == 'command' || type == 'commands'
- then by key alphabetically

**Comment by macintacos** [comment]
@stevenguh I'll look into this tonight and let you know how things are going/if I hit any snags!

**Comment by macintacos** [comment]
@stevenguh just so that we're clear, after taking a look at what Spacemacs does, it seems like it prefers the following order:

1. Symbols/numbers (I can't really tell what order here?)
2. Lowercase commands (transient/single commands, doesn't matter)
3. Uppercase commands (transient/single commands, doesn't matter)
4. Menus (symbols, lowercase, then uppercase)

The above order is what I'm tailoring the sorting function to.

**Comment by stevenguh** [comment]
Let's try that. One thing I am thinking is that the top level menu is going to full of numbers before we see useful items, but we will have a consistent order.

**Comment by macintacos** [comment]
Yeah it'd be nice if we could show something like `1..9` like Spacemacs, but I'm not sure if that's possible using the current setup.

Opened #132 to sort everything 🙂

**Comment by stevenguh** [comment]
It would be nice, but not possible today. However, my current spacemacs shows both `1...9` and as well as individual numbers. Wired. Well I guess we are consistent😄
<img width="898" alt="image" src="https://user-images.githubusercontent.com/3455662/96833680-b2a4eb80-13f5-11eb-8948-69d74b50caa0.png">

**Comment by stevenguh** [comment]
Thank you @gnomesoup and @macintacos 😁
